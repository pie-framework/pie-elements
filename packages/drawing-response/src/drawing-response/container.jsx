import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';

import constants from './constants';
import Button from './button';
import DrawablePalette from './drawable-palette';
import DrawableMain from './drawable-main';
import DrawableText from './drawable-text';
import Icon from './icon';

const { tools: TOOLS } = constants;

const ROGVAIV = [
  'red',
  'orange',
  'yellow',
  'violet',
  'blue',
  'green',
  'white',
  'black'
].map(c => ({ value: c, label: c }));

export class Container extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    disabled: PropTypes.bool,
    session: PropTypes.object.isRequired,
    onSessionChange: PropTypes.func.isRequired,
    imageDimensions: PropTypes.object.isRequired,
    imageUrl: PropTypes.string.isRequired
  };

  constructor(props) {
    super(props);
    const TextEntry = new DrawableText();

    this.state = {
      drawableDimensions: {
        height: 0,
        width: 0
      },
      toolActive: TOOLS[0],
      fillColor: 'white',
      fillColorList: [
        { value: 'transparent', label: 'no fill' },
        { value: 'lightblue', label: 'lightblue' },
        { value: 'lightyellow', label: 'lightyellow' },
        ...ROGVAIV
      ],
      outlineColor: 'black',
      outlineColorList: ROGVAIV,
      paintColor: 'red',
      paintColorList: ROGVAIV,
      TextEntry
    };
  }

  setDimensions() {
    const checkExist = setInterval(() => {
      try {
        const { height, width } = this.drawable.getBoundingClientRect();
        if (height !== 0 && width !== 0) {
          this.setState({
            drawableDimensions: {
              height,
              width
            }
          });
        }
      } catch (e) {
        // eslint-disable-next-line no-console
        console.warn('setDimensions Error: ', e);
      } finally {
        clearInterval(checkExist);
      }
    }, 100);
  }

  componentDidMount() {
    this.setDimensions();
  }

  componentWillUnmount() {
    const { TextEntry } = this.state;

    TextEntry.removeEventListeners();
  }

  handleMakeToolActive(tool) {
    const { TextEntry } = this.state;
    const { type } = tool;

    if (type !== 'Text') {
      this.setState({
        toolActive: tool
      });
    } else {
      TextEntry.addNewTextEntry();
      // Force update
      this.setState({
        updatedAt: new Date()
      });
    }
  }

  checkIfToolIsDisabled = type => {
    if (this.props.disabled) return true;


    const { toolActive } = this.state;
    // Text will never be disabled since on each "Text Entry" click a new text is added
    if (type === 'Text') {
      return false;
    }
    return type === toolActive.type;
  };

  handleColorChange(type, color) {
    const cType = `${type}Color`;
    this.setState({
      [cType]: color
    });
  }

  render() {
    const {
      classes,
      disabled,
      imageUrl,
      imageDimensions,
      onSessionChange,
      session
    } = this.props;
    const {
      drawableDimensions,
      toolActive,
      fillColor,
      fillColorList,
      outlineColor,
      outlineColorList,
      paintColor,
      paintColorList,
      TextEntry
    } = this.state;

    return (
      <div className={classes.base}>
        <DrawablePalette
          fillColor={fillColor}
          fillList={fillColorList}
          outlineColor={outlineColor}
          outlineList={outlineColorList}
          paintColor={paintColor}
          paintList={paintColorList}
          onFillColorChange={color => this.handleColorChange('fill', color)}
          onOutlineColorChange={color =>
            this.handleColorChange('outline', color)
          }
          onPaintColorChange={color => this.handleColorChange('paint', color)}
        />

        <div className={classes.box}>
          <div className={classes.toolbar}>
            <div>
              {TOOLS.map(tool => {
                const { type, label, icon } = tool;

                return (
                  <Button
                    title={label}
                    key={type}
                    disabled={this.checkIfToolIsDisabled(type)}
                    onClick={() => this.handleMakeToolActive(tool)}
                    label={<Icon path={icon} />}
                  />
                );
              })}
            </div>
          </div>

          <div
            ref={drawable => {
              this.drawable = drawable;
            }}
            className={classes.drawableHeight}
          >
            <DrawableMain
              session={session}
              disabled={disabled}
              onSessionChange={onSessionChange}
              fillColor={fillColor}
              outlineColor={outlineColor}
              paintColor={paintColor}
              imageUrl={imageUrl}
              drawableDimensions={drawableDimensions}
              imageDimensions={imageDimensions}
              toolActive={toolActive}
              TextEntry={TextEntry}
            />
          </div>
        </div>
      </div>
    );
  }
}

const styles = theme => ({
  base: {
    marginTop: theme.spacing.unit * 2
  },
  box: {
    border: '1px solid #E0E1E6',
    borderRadius: '5px',
    marginTop: theme.spacing.unit * 2
  },
  drawableHeight: {
    minHeight: 350
  },
  toolbar: {
    backgroundColor: '#ECEDF1',
    borderBottom: '1px solid #E0E1E6',
    borderTopLeftRadius: '5px',
    borderTopRightRadius: '5px',
    display: 'flex',
    justifyContent: 'space-between',
    padding: '12px 8px'
  }
});

export default withStyles(styles)(Container);
