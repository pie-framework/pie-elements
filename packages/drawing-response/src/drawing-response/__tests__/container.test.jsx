import { shallow } from 'enzyme';
import React from 'react';

global.MutationObserver = class {
  constructor(callback) {}
  disconnect() {}
  observe(element, initObject) {}
};

import { Container } from '../container';

beforeEach(() => {
  jest.useFakeTimers();
});
describe('Container', () => {
  let w;
  let onSessionChange = jest.fn();
  const wrapper = (extras, renderOpts) => {
    const defaults = {
      classes: {},
      className: 'className',
      onSessionChange,
      imageDimensions: {},
      imageUrl: 'url',
      session: {}
    };
    const props = { ...defaults, ...extras };
    return shallow(<Container {...props} />, {
      disableLifecycleMethods: false,
      ...renderOpts
    });
  };

  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });

    it('renders disabled', () => {
      w = wrapper({ disabled: true });
      expect(w).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    describe('setDimensions', () => {
      it('handles errors and clears interval', () => {
        const w = wrapper({}, { disableLifecycleMethods: true });
        w.instance().drawable = undefined;
        w.instance().setDimensions();
        jest.runAllTimers();
        expect(clearInterval).toHaveBeenCalled();
      });
    });
  });
});
