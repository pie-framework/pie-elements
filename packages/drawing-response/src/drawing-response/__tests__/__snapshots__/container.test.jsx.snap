// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container snapshot renders 1`] = `
<div>
  <WithStyles(Palette)
    fillColor="white"
    fillList={
      Array [
        Object {
          "label": "no fill",
          "value": "transparent",
        },
        Object {
          "label": "lightblue",
          "value": "lightblue",
        },
        Object {
          "label": "lightyellow",
          "value": "lightyellow",
        },
        Object {
          "label": "red",
          "value": "red",
        },
        Object {
          "label": "orange",
          "value": "orange",
        },
        Object {
          "label": "yellow",
          "value": "yellow",
        },
        Object {
          "label": "violet",
          "value": "violet",
        },
        Object {
          "label": "blue",
          "value": "blue",
        },
        Object {
          "label": "green",
          "value": "green",
        },
        Object {
          "label": "white",
          "value": "white",
        },
        Object {
          "label": "black",
          "value": "black",
        },
      ]
    }
    onFillColorChange={[Function]}
    onOutlineColorChange={[Function]}
    onPaintColorChange={[Function]}
    outlineColor="black"
    outlineList={
      Array [
        Object {
          "label": "red",
          "value": "red",
        },
        Object {
          "label": "orange",
          "value": "orange",
        },
        Object {
          "label": "yellow",
          "value": "yellow",
        },
        Object {
          "label": "violet",
          "value": "violet",
        },
        Object {
          "label": "blue",
          "value": "blue",
        },
        Object {
          "label": "green",
          "value": "green",
        },
        Object {
          "label": "white",
          "value": "white",
        },
        Object {
          "label": "black",
          "value": "black",
        },
      ]
    }
    paintColor="red"
    paintList={
      Array [
        Object {
          "label": "red",
          "value": "red",
        },
        Object {
          "label": "orange",
          "value": "orange",
        },
        Object {
          "label": "yellow",
          "value": "yellow",
        },
        Object {
          "label": "violet",
          "value": "violet",
        },
        Object {
          "label": "blue",
          "value": "blue",
        },
        Object {
          "label": "green",
          "value": "green",
        },
        Object {
          "label": "white",
          "value": "white",
        },
        Object {
          "label": "black",
          "value": "black",
        },
      ]
    }
  />
  <div>
    <div>
      <WithStyles(RawButton)
        disabled={true}
        key="Select"
        label={
          <IconWrapper
            path="mdiCursorDefault"
          />
        }
        onClick={[Function]}
        title="Select"
      />
      <WithStyles(RawButton)
        disabled={false}
        key="FreePathDrawable"
        label={
          <IconWrapper
            path="mdiPencil"
          />
        }
        onClick={[Function]}
        title="Free Draw"
      />
      <WithStyles(RawButton)
        disabled={false}
        key="LineDrawable"
        label={
          <IconWrapper
            path="mdiMinus"
          />
        }
        onClick={[Function]}
        title="Line"
      />
      <WithStyles(RawButton)
        disabled={false}
        key="RectangleDrawable"
        label={
          <IconWrapper
            path="mdiRectangle"
          />
        }
        onClick={[Function]}
        title="Rectangle"
      />
      <WithStyles(RawButton)
        disabled={false}
        key="CircleDrawable"
        label={
          <IconWrapper
            path="mdiCircle"
          />
        }
        onClick={[Function]}
        title="Circle"
      />
      <WithStyles(RawButton)
        disabled={false}
        key="Text"
        label={
          <IconWrapper
            path="mdiFormatColorText"
          />
        }
        onClick={[Function]}
        title="Text Entry"
      />
      <WithStyles(RawButton)
        disabled={false}
        key="EraserDrawable"
        label={
          <IconWrapper
            path="mdiEraser"
          />
        }
        onClick={[Function]}
        title="Eraser"
      />
    </div>
    <div
      style={
        Object {
          "height": 0,
          "maxHeight": 0,
        }
      }
    >
      <WithStyles(DrawableMain)
        TextEntry={
          TextDrawable {
            "addNewTextEntry": [Function],
            "all": Array [],
            "eventListenersDetachArray": Array [],
            "handleClick": [Function],
            "handleDblClick": [Function],
            "handleMouseDown": [Function],
            "handleMouseUp": [Function],
            "handleTransform": [Function],
            "removeEventListeners": [Function],
            "setAll": [Function],
          }
        }
        drawableDimensions={
          Object {
            "height": 0,
            "width": 0,
          }
        }
        fillColor="white"
        imageDimensions={Object {}}
        imageUrl="url"
        onSessionChange={[MockFunction]}
        outlineColor="black"
        paintColor="red"
        scale={1}
        session={Object {}}
        toolActive={
          Object {
            "icon": "mdiCursorDefault",
            "label": "Select",
            "type": "Select",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Container snapshot renders disabled 1`] = `
<div>
  <div>
    <div>
      <WithStyles(RawButton)
        disabled={true}
        key="Select"
        label={
          <IconWrapper
            path="mdiCursorDefault"
          />
        }
        onClick={[Function]}
        title="Select"
      />
      <WithStyles(RawButton)
        disabled={true}
        key="FreePathDrawable"
        label={
          <IconWrapper
            path="mdiPencil"
          />
        }
        onClick={[Function]}
        title="Free Draw"
      />
      <WithStyles(RawButton)
        disabled={true}
        key="LineDrawable"
        label={
          <IconWrapper
            path="mdiMinus"
          />
        }
        onClick={[Function]}
        title="Line"
      />
      <WithStyles(RawButton)
        disabled={true}
        key="RectangleDrawable"
        label={
          <IconWrapper
            path="mdiRectangle"
          />
        }
        onClick={[Function]}
        title="Rectangle"
      />
      <WithStyles(RawButton)
        disabled={true}
        key="CircleDrawable"
        label={
          <IconWrapper
            path="mdiCircle"
          />
        }
        onClick={[Function]}
        title="Circle"
      />
      <WithStyles(RawButton)
        disabled={true}
        key="Text"
        label={
          <IconWrapper
            path="mdiFormatColorText"
          />
        }
        onClick={[Function]}
        title="Text Entry"
      />
      <WithStyles(RawButton)
        disabled={true}
        key="EraserDrawable"
        label={
          <IconWrapper
            path="mdiEraser"
          />
        }
        onClick={[Function]}
        title="Eraser"
      />
    </div>
    <div
      style={
        Object {
          "height": 0,
          "maxHeight": 0,
        }
      }
    >
      <WithStyles(DrawableMain)
        TextEntry={
          TextDrawable {
            "addNewTextEntry": [Function],
            "all": Array [],
            "eventListenersDetachArray": Array [],
            "handleClick": [Function],
            "handleDblClick": [Function],
            "handleMouseDown": [Function],
            "handleMouseUp": [Function],
            "handleTransform": [Function],
            "removeEventListeners": [Function],
            "setAll": [Function],
          }
        }
        disabled={true}
        drawableDimensions={
          Object {
            "height": 0,
            "width": 0,
          }
        }
        fillColor="white"
        imageDimensions={Object {}}
        imageUrl="url"
        onSessionChange={[MockFunction]}
        outlineColor="black"
        paintColor="red"
        scale={1}
        session={Object {}}
        toolActive={
          Object {
            "icon": "mdiCursorDefault",
            "label": "Select",
            "type": "Select",
          }
        }
      />
    </div>
  </div>
</div>
`;
